python/packages/azurefunctions/agent_framework_azurefunctions/_durable_agent_state.py

```diff
@@ -0,0 +1,1242 @@
+        """
+        self.data = DurableAgentStateData()
+        self.schema_version = schema_version
+
+    def to_dict(self) -> dict[str, Any]:
+        # Serialize conversationHistory
+        serialized_history: list[dict[str, Any]] = []
+        for entry in self.data.conversationHistory:
+            # Properly serialize each entry to a dictionary
+            if hasattr(entry, "to_dict"):
+                serialized_history.append(entry.to_dict())
+            else:
+                # Fallback for already-serialized entries
+                serialized_history.append(entry)  # type: ignore
+
+        return {
+            "schemaVersion": self.schema_version,
+            "data": {"conversationHistory": serialized_history},
+        }
+
+    def to_json(self) -> str:
+        return json.dumps(self.to_dict())
+
+    @classmethod
+    def from_dict(cls, obj: dict[str, Any]) -> DurableAgentState:
+        schema_version = obj.get("schemaVersion")
+        if not schema_version:
+            raise ValueError("The durable agent state is missing the 'schemaVersion' property.")
+
+        if not schema_version.startswith("1."):
+            raise ValueError(f"The durable agent state schema version '{schema_version}' is not supported.")
+
+        data_dict = obj.get("data")
+        if data_dict is None:
+            raise ValueError("The durable agent state is missing the 'data' property.")
+
+        instance = cls(schema_version=schema_version)
+        # Deserialize the data dict into DurableAgentStateData
+        if isinstance(data_dict, dict):
+            instance.data = DurableAgentStateData(
+                conversationHistory=data_dict.get("conversationHistory", []),
+                extensionData=data_dict.get("extensionData"),
+            )
+        return instance
+
+    @classmethod
+    def from_json(cls, json_str: str) -> DurableAgentState:
+        try:
+            obj = json.loads(json_str)
+        except json.JSONDecodeError as e:
+            raise ValueError("The durable agent state is not valid JSON.") from e
+
+        return cls.from_dict(obj)
+
+    def restore_state(self, state: dict[str, Any]) -> None:
+        """Restore state from a dictionary.
+
+        Args:
+            state: Dictionary containing schemaVersion and data (full state structure)
+        """
+        # Extract the data portion from the state
+        data_dict = state.get("data", {})
+
+        # Restore the conversation history - deserialize entries from dicts to objects
+        history_data = data_dict.get("conversationHistory", [])
+        deserialized_history: list[DurableAgentStateEntry] = []
+        for entry_dict in history_data:
+            if isinstance(entry_dict, dict):
+                # Deserialize based on $type discriminator
+                entry_type = entry_dict.get("$type") or entry_dict.get("json_type")
+                if entry_type == "response":
```

Comment:
NIT: can we make these class constants or something? Maybe an enums `DurableStateConstants` or something for cleanliness?